src/orderbook/state.ts

```diff
@@ -28,7 +83,139 @@ export class AccountState {
       )
     }
     this.batchId = batchId
-    // TODO(nlordell): Actually apply events to state.
+
+    for (const ev of events) {
+      const data = ev.returnValues
+      switch (ev.event) {
+        case 'OrderCancellation':
+          this.order(data.owner, parseInt(data.id)).validUntil = (await this.batches.forEvent(ev)) - 1
+          break
+        case 'OrderDeletion':
+          this.deleteOrder(data.owner, parseInt(data.id))
+          break
+        case 'OrderPlacement':
+          this.addOrder(data.owner, {
+            buyToken: parseInt(data.buyToken),
+            sellToken: parseInt(data.sellToken),
+            validFrom: parseInt(data.validFrom),
+            validUntil: parseInt(data.validUntil),
+            priceNumerator: BigInt(data.priceNumerator),
+            priceDenominator: BigInt(data.priceDenominator),
+          })
+          break
+        case 'TokenListing':
+          this.addToken(parseInt(data.id), data.token)
+          break
+        case 'Deposit':
+        case 'SolutionSubmission':
+        case 'Trade':
+        case 'TradeReversion':
+        case 'WithdrawRequest':
+        case 'Withdraw':
+          // TODO(nlordell): Handle more events.
+          break
+        default:
+          throw new UnhandledEventError(ev)
+      }
+    }
+  }
+
+  /**
+   * Adds a token to the account state.
+   *
+   * @throws If a token with the same ID has already been added.
+   */
+  private addToken(id: TokenId, addr: Address): void {
+    assert(!this.tokens.has(id), `token ${id} at ${addr} added more than once`)
+    this.tokens.set(id, addr)
+  }
+
+  /**
+   * Gets the account data for the specified user, creating an empty one if it
+   * does not already exist.
+   */
+  private account(user: Address): Account {
+    let account = this.accounts.get(user)
+    if (account === undefined) {
+      account = {
+        orders: [],
+      }
+      this.accounts.set(user, account)
+    }
+
+    return account
+  }
+
+  /**
+   * Adds a user order and returns the order ID of the newly created order.
+   */
+  private addOrder(user: Address, order: Order): number {
+    const userOrders = this.account(user).orders
```

Comment:
Should we assert here that the id is actually the same as given in the event data?